### PR TITLE
tidbmonitor add readiness probe

### LINKED ISSUE #3935: Support setting liveness probe and readiness probe for TidbMonitor
## Feature Request

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

Support setting liveness probe and readiness probe for Grafana and Prometheus containers in TidbMonitor.
**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by mikechengwei**
### What problem does this PR solve?
Fix #3935 

### What is changed and how does it work?
Tidbmonitor add readiness probe.

### Code changes

- [ ] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [ ] E2E test 
- [ ] Manual test 
- [ ] No code 

### Side effects

Rolling update tidbmonitor pod.

### Release Notes

```release-note
Add readiness probe for the monitor Pod. This will cause the monitor Pod to be recreated
```

### CODE REVIEW

**Review by july2993 — APPROVED** [review]

**Review by july2993 — COMMENTED** [review]
Is there any knows issue without the probe before this PR?

**Comment by mikechengwei** [comment]
> Is there any knows issue without the probe before this PR?
I think there should be none, this pr is an optimization operation.

File: pkg/monitor/monitor/util.go [hunk 1]
```diff
@@ -491,6 +491,25 @@ func getMonitorPrometheusContainer(monitor *v1alpha1.TidbMonitor, tc *v1alpha1.T
...
+	var readinessProbeHandler core.Handler
+	{
+		readyPath := "/-/ready"
+		readinessProbeHandler.HTTPGet = &core.HTTPGetAction{
+			Path: readyPath,
+			Port: intstr.FromInt(9090),
+		}
+
+	}
+	readinessProbe := &core.Probe{
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/util.go:505]: Please also add livenessProbe.

Comment (DanielZhangQD) [reply pkg/monitor/monitor/util.go:505]: And what about also adding the probes for the Grafana.

Comment (mikechengwei) [reply pkg/monitor/monitor/util.go:505]: I don’t recommend adding LivenessProbe, you can refer to https://github.com/prometheus-operator/prometheus-operator/pull/3502.

Comment (DanielZhangQD) [reply pkg/monitor/monitor/util.go:505]: OK.

File: pkg/monitor/monitor/util.go [hunk 2]
```diff
@@ -555,6 +574,35 @@ func getMonitorGrafanaContainer(secret *core.Secret, monitor *v1alpha1.TidbMonit
...
+		}
+
+	}
+
+	readinessProbe := &core.Probe{
+		Handler:             probeHandler,
+		TimeoutSeconds:      30,
+		PeriodSeconds:       10,
+		SuccessThreshold:    1,
+		InitialDelaySeconds: 30,
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/util.go:594]: Move this to the liveness probe?

File: pkg/monitor/monitor/util.go [hunk 3]
```diff
@@ -555,6 +574,35 @@ func getMonitorGrafanaContainer(secret *core.Secret, monitor *v1alpha1.TidbMonit
...
+	}
+	livenessProbe := &core.Probe{
+		Handler:          probeHandler,
+		TimeoutSeconds:   1,
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/util.go:598]: Why this is different from the value for the readiness probe?

Comment (DanielZhangQD) [reply pkg/monitor/monitor/util.go:598]: How are the values determined?

File: pkg/monitor/monitor/util.go [hunk 4]
```diff
@@ -479,6 +479,25 @@ func getMonitorPrometheusContainer(monitor *v1alpha1.TidbMonitor, tc *v1alpha1.T
...
+		}
+
+	}
+	readinessProbe := &core.Probe{
+		Handler:          readinessProbeHandler,
+		TimeoutSeconds:   3,
+		PeriodSeconds:    5,
+		FailureThreshold: 120, // Allow up to 10m on startup for data recovery
+	}
+	c.ReadinessProbe = readinessProbe
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/util.go:499]: Add a comment for the liveness probe here?

**Review by DanielZhangQD — APPROVED** [review]
LGTM

File: pkg/monitor/monitor/util.go [hunk 5]
```diff
@@ -479,6 +479,25 @@ func getMonitorPrometheusContainer(monitor *v1alpha1.TidbMonitor, tc *v1alpha1.T
 		commands = append(commands, "--storage.tsdb.max-block-duration=2h")
 		commands = append(commands, "--storage.tsdb.min-block-duration=2h")
 	}
+
+	//Add readiness probe. LivenessProbe probe will affect prom wal replay,ref :prometheus-operator/prometheus-operator#3502
```
Comment (csuzhangxc) [inline pkg/monitor/monitor/util.go:483]:
```suggestion
	//Add readiness probe. LivenessProbe probe will affect prom wal replay,ref: https://github.com/prometheus-operator/prometheus-operator/pull/3502
```

**Review by csuzhangxc — APPROVED** [review]
LGTM

**Pull request merged**